File: projects/native-apps/ios/Shared/SharedArticlesDrop.swift
import SwiftUI

enum SharedArticlesDrop: Hashable {
	case always(label: String)
	case ticked(Readlist)
	case unticked(Readlist)

	init(readlist: Readlist, mainlineHref: String?, tickedHrefs: Set<String>) {
		if readlist.href == mainlineHref {
			self = .always(label: readlist.label)
		} else if tickedHrefs.contains(readlist.href) {
			self = .ticked(readlist)
		} else {
			self = .unticked(readlist)
		}
	}

	static func firstAsk(readlists: [Readlist], mainlineHref: String?) -> [SharedArticlesDrop] {
		readlists.map { SharedArticlesDrop(readlist: $0, mainlineHref: mainlineHref, tickedHrefs: []) }
	}

	var label: String {
		switch self {
		case .always(let label): return label
		case .ticked(let readlist), .unticked(let readlist): return readlist.label
		}
	}

	var choice: Readlist? {
		switch self {
		case .always: return nil
		case .ticked(let readlist), .unticked(let readlist): return readlist
		}
	}

	var showsTick: Bool {
		switch self {
		case .always, .ticked: return true
		case .unticked: return false
		}
	}
}

enum SharedArticlesDropPresentation {
	static func title(for drop: SharedArticlesDrop) -> String {
		drop.showsTick ? "Shared articles drop here" : "Want shared article to drop here?"
	}

	static func boxSystemImage(for drop: SharedArticlesDrop) -> String {
		drop.showsTick ? "checkmark.square.fill" : "square"
	}

	static func boxTint(for drop: SharedArticlesDrop) -> Color {
		drop.showsTick ? .brandSuccessText : .brandTextSecondary
	}

	static func titleTint(for drop: SharedArticlesDrop) -> Color {
		drop.showsTick ? .brandTextPrimary : .brandTextSecondary
	}

	static func lockSystemImage(for drop: SharedArticlesDrop) -> String? {
		drop.choice == nil ? "lock.fill" : nil
	}

	static func borderOpacity(for drop: SharedArticlesDrop) -> Double {
		drop.showsTick ? 1 : 0.35
	}

	static func accessibilityTraits(for drop: SharedArticlesDrop) -> AccessibilityTraits {
		drop.showsTick ? [.isButton, .isSelected] : .isButton
	}

	static func explanation(for drop: SharedArticlesDrop) -> String? {
		guard drop.choice == nil else { return nil }
		return "Articles always drop on \(drop.label). Reading one marks as read in all readlists."
	}
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/core/SharedArticlesDrop.kt
package com.readplace.android.core

/**
 * One readlist's row in the "where do shared articles drop?" surfaces (the list's
 * drop row and the share chooser). Mainline is [Always] — locked and checked,
 * because every save drops there — while an extra readlist is [Ticked] or
 * [Unticked] by whether the reader chose it. Mainline is identified only by the
 * advertised root href; a readlist is never guessed to be mainline from its label
 * or position.
 */
sealed interface SharedArticlesDrop {
	data class Always(override val label: String) : SharedArticlesDrop

	data class Ticked(val readlist: Readlist) : SharedArticlesDrop {
		override val label: String get() = readlist.label
	}

	data class Unticked(val readlist: Readlist) : SharedArticlesDrop {
		override val label: String get() = readlist.label
	}

	val label: String

	/** The readlist this row toggles, or null for locked mainline. */
	val choice: Readlist? get() = when (this) {
		is Always -> null
		is Ticked -> readlist
		is Unticked -> readlist
	}

	/** Whether the row shows a tick — always for mainline and a chosen extra. */
	val showsTick: Boolean get() = when (this) {
		is Always, is Ticked -> true
		is Unticked -> false
	}

	companion object {
		fun of(readlist: Readlist, mainlineHref: String?, tickedHrefs: Set<String>): SharedArticlesDrop = when {
			readlist.href == mainlineHref -> Always(label = readlist.label)
			tickedHrefs.contains(readlist.href) -> Ticked(readlist)
			else -> Unticked(readlist)
		}

		/** The rows for a first, never-answered ask: every extra starts unticked. */
		fun firstAsk(readlists: List<Readlist>, mainlineHref: String?): List<SharedArticlesDrop> =
			readlists.map { of(it, mainlineHref, tickedHrefs = emptySet()) }
	}
}

/** The reader-facing copy for a [SharedArticlesDrop], kept out of the Compose layer
 * so the strings are asserted in unit tests. Icons, tints and touch targets are the
 * view's own presentation, derived there from [SharedArticlesDrop.showsTick] and
 * [SharedArticlesDrop.choice]. */
object SharedArticlesDropPresentation {
	fun title(drop: SharedArticlesDrop): String =
		if (drop.showsTick) "Shared articles drop here" else "Want shared article to drop here?"

	/** The explanation shown for locked mainline (and only mainline): articles
	 * always land there and reading one marks it read across every readlist. Null
	 * for a toggleable extra, which needs none. */
	fun explanation(drop: SharedArticlesDrop): String? =
		if (drop.choice == null) {
			"Articles always drop on ${drop.label}. Reading one marks as read in all readlists."
		} else {
			null
		}
}
